src/OrbitGl/App.cpp

```diff
@@ -2020,7 +2002,117 @@ orbit_base::Future<ErrorMessageOr<CanceledOr<void>>> OrbitApp::RetrieveModuleAnd
   return load_result;
 }
 
-orbit_base::Future<ErrorMessageOr<CanceledOr<std::filesystem::path>>> OrbitApp::RetrieveModule(
+Future<ErrorMessageOr<CanceledOr<std::filesystem::path>>> OrbitApp::RetrieveModuleViaDownload(
+    const ModuleIdentifier& module_id) {
+  if (const auto it = symbol_files_currently_downloading_.find(module_id.file_path);
+      it != symbol_files_currently_downloading_.end()) {
+    return it->second.future;
+  }
+
+  orbit_base::StopSource stop_source;
+
+  using SymbolRetrieveResult = ErrorMessageOr<CanceledOr<std::filesystem::path>>;
+  Future<SymbolRetrieveResult> retrieve_from_instance_future =
+      orbit_base::UnwrapFuture(main_thread_executor_->Schedule(
+          [this, module_id,
+           stop_token = stop_source.GetStopToken()]() mutable -> Future<SymbolRetrieveResult> {
+            // If --local, Orbit cannot download files from the instance, because no ssh channel
+            // exists. We still return an ErrorMessage to enable continuing searching for symbols
+            // from other symbol sources.
+            if (absl::GetFlag(FLAGS_local) || !main_window_->IsConnected()) {
+              return {ErrorMessage{"\n- Not able to search for symbols from instance"}};
+            }
+
+            return RetrieveModuleFromRemote(module_id.file_path, std::move(stop_token))
+                .Then(main_thread_executor_,
+                      [module_id](const SymbolRetrieveResult& retrieve_result) mutable
+                      -> SymbolRetrieveResult {
+                        if (retrieve_result.has_value()) return retrieve_result;
+
+                        return ErrorMessage{
+                            absl::StrFormat("\n- Did not find symbols from instance: %s",
+                                            retrieve_result.error().message())};
+                      });
+          }));
```

Comment:
I think this does not need to be wrapped into another `main_thread_executer_->Schedule`. `RetrieveModuleViaDownload` Should only be called from the main thread anyways, so this could look something like this:
```
  // If --local, Orbit cannot download files from the instance, because no ssh channel exists. We
  // still return an ErrorMessage to enable continuing searching for symbols from other symbol
  // sources.
  if (absl::GetFlag(FLAGS_local) || !main_window_->IsConnected()) {
    return {ErrorMessage{"\n- Not able to search for symbols from instance"}};
  }

  orbit_base::StopSource stop_source;

  using SymbolRetrieveResult = ErrorMessageOr<CanceledOr<std::filesystem::path>>;
  Future<SymbolRetrieveResult> retrieve_from_instance_future =
      RetrieveModuleFromRemote(module_id.file_path, stop_source.GetStopToken())
          .Then(main_thread_executor_,
                [module_id](
                    const SymbolRetrieveResult& retrieve_result) mutable -> SymbolRetrieveResult {
                  if (retrieve_result.has_value()) return retrieve_result;

                  return ErrorMessage{absl::StrFormat("\n- Did not find symbols from instance: %s",
                                                      retrieve_result.error().message())};
                });
```

Reply:
I just read and tried more of the code and its not as easy as I thought, because then we dont try the symbol stores when we are in `--local`. You could either leave the code as is or do something like: 
```
  using SymbolRetrieveResult = ErrorMessageOr<CanceledOr<std::filesystem::path>>;
  Future<SymbolRetrieveResult> retrieve_from_instance_future{ErrorMessage{"\n- Not able to search for symbols from instance"}};

  if (!absl::GetFlag(FLAGS_local) && main_window_->IsConnected()) {
    retrieve_from_instance_future =
        RetrieveModuleFromRemote(module_id.file_path, stop_source.GetStopToken())
            .Then(
                main_thread_executor_,
                [module_id](
                    const SymbolRetrieveResult& retrieve_result) mutable -> SymbolRetrieveResult {
                  if (retrieve_result.has_value()) return retrieve_result;

                  return ErrorMessage{absl::StrFormat("\n- Did not find symbols from instance: %s",
                                                      retrieve_result.error().message())};
                });
  }
```